### PR TITLE
Deprecate nil_typet [DOC-12]

### PR DESCRIPTION
**Pull request by majakusber**
This is part of the `typet` cleanup, based on top of https://github.com/diffblue/cbmc/pull/2815, please review only the last commit.

### CODE REVIEW

**Review by thk123 — APPROVED** [review]

**Comment by thk123** [comment]
Don't merge until #2815 is merged.

**Review by mgudemann — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 426debe22737fadd5a163578b6ba7ff518e6986c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84533893

**Comment by majakusber** [comment]
diffblue/cbmc#2815 is merged now, rebased on develop.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 4f03876b4ba5400c6b2c1238583532a46cace0f4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85007182
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 03740189647b1930f077ff49f0ee7176cfe8dba6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85011006

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
